internal/app/machined/main.go

```diff
@@ -80,217 +79,90 @@ func sync() {
 	log.Printf("sync hasn't completed in time, aborting...")
 }
 
-func init() {
-	// Explicitly set the default http client transport
-	// to work around our fun proxy.Do once bug.
-	// This is the http.DefaultTransport with the Proxy
-	// func overridden so that the environment variables
-	// with be reread/initialized each time the http call
-	// is made.
-	http.DefaultClient.Transport = &http.Transport{
-		Proxy: func(req *http.Request) (*url.URL, error) {
-			return httpproxy.FromEnvironment().ProxyFunc()(req.URL)
-		},
-		DialContext: (&net.Dialer{
-			Timeout:   30 * time.Second,
-			KeepAlive: 30 * time.Second,
-			DualStack: true,
-		}).DialContext,
-		ForceAttemptHTTP2:     true,
-		MaxIdleConns:          100,
-		IdleConnTimeout:       90 * time.Second,
-		TLSHandshakeTimeout:   10 * time.Second,
-		ExpectContinueTimeout: 1 * time.Second,
-	}
-}
-
-// nolint: gocyclo
-func revert() (err error) {
-	f, err := os.OpenFile(syslinux.SyslinuxLdlinux, os.O_RDWR, 0700)
-	if err != nil {
-		if errors.Is(err, os.ErrNotExist) {
-			return nil
-		}
+func reboot(err error) {
+	log.Print(err)
 
-		return err
+	for i := 10; i >= 0; i-- {
+		log.Printf("rebooting in %d seconds\n", i)
+		time.Sleep(1 * time.Second)
 	}
 
-	// nolint: errcheck
-	defer f.Close()
-
-	adv, err := syslinux.NewADV(f)
-	if err != nil {
-		return err
-	}
-
-	label, ok := adv.ReadTag(syslinux.AdvUpgrade)
-	if !ok {
-		return nil
-	}
-
-	if label == "" {
-		adv.DeleteTag(syslinux.AdvUpgrade)
-
-		if _, err = f.Write(adv); err != nil {
-			return err
-		}
-
-		return nil
-	}
-
-	log.Printf("reverting default boot to %q", label)
-
-	var b []byte
-
-	if b, err = ioutil.ReadFile(syslinux.SyslinuxConfig); err != nil {
-		return err
-	}
-
-	re := regexp.MustCompile(`^DEFAULT\s(.*)`)
-	matches := re.FindSubmatch(b)
-
-	if len(matches) != 2 {
-		return fmt.Errorf("expected 2 matches, got %d", len(matches))
-	}
-
-	b = re.ReplaceAll(b, []byte(fmt.Sprintf("DEFAULT %s", label)))
-
-	if err = ioutil.WriteFile(syslinux.SyslinuxConfig, b, 0600); err != nil {
-		return err
-	}
-
-	adv.DeleteTag(syslinux.AdvUpgrade)
-
-	if _, err = f.Write(adv); err != nil {
-		return err
-	}
+	sync()
 
-	return nil
+	// nolint: errcheck
+	unix.Reboot(unix.LINUX_REBOOT_CMD_RESTART)
 }
 
-// nolint: gocyclo
 func main() {
-	var err error
-
-	// This is main entrypoint into machined execution, control is passed here
-	// from init after switch root.
-	//
-	// When machined terminates either on normal shutdown (reboot, poweroff), or
-	// due to panic, control goes through recovery() and reboot() functions
-	// below, which finalize node state - sync buffers, initiate poweroff or
-	// reboot. Also on shutdown, other deferred function are called, for example
-	// services are gracefully shutdown.
-
-	// On any return from init.main(), initiate host reboot or shutdown handle
-	// any panics in the main goroutine, and proceed to reboot() above
-	defer recovery()
```

Comment:
I think we should catch/recover panics here. I know it's not perfect (doesn't cover all the goroutines), but better than nothing, as init panic is going to lead to kernel panic.

Reply:
Fixed